libafl_qemu/src/modules/usermode/redirect_stdin.rs

```diff
@@ -0,0 +1,140 @@
+use core::fmt::Debug;
+
+use libafl_bolts::HasLen;
+use libafl_qemu_sys::GuestAddr;
+
+#[cfg(not(cpu_target = "hexagon"))]
+use crate::SYS_read;
+use crate::{
+    Qemu,
+    emu::EmulatorModules,
+    modules::{EmulatorModule, EmulatorModuleTuple},
+    qemu::{Hook, SyscallHookResult},
+};
+
+#[cfg(cpu_target = "hexagon")]
+/// Hexagon syscalls are not currently supported by the `syscalls` crate, so we just paste this here for now.
+/// <https://github.com/qemu/qemu/blob/11be70677c70fdccd452a3233653949b79e97908/linux-user/hexagon/syscall_nr.h#L230>
+#[expect(non_upper_case_globals)]
+const SYS_read: u8 = 63;
+
+/// This module hijacks any read to buffer from stdin, and instead fill the buffer from the specified input address
+/// This is useful when your binary target reads the input from the stdin.
+/// With this you can just fuzz more like afl++
+/// You need to use this with snapshot module!
+#[derive(Debug, Default)]
+pub struct RedirectStdinModule {
+    input_addr: Option<GuestAddr>,
+    read: usize,
+    total: usize,
+}
+
+impl RedirectStdinModule {
+    #[must_use]
+    /// constuctor
+    pub fn new() -> Self {
+        Self::with_input_addr(None)
+    }
+
+    #[must_use]
+    /// Create with specified input address
+    pub fn with_input_addr(addr: Option<GuestAddr>) -> Self {
+        Self {
+            input_addr: addr,
+            read: 0,
+            total: 0,
+        }
+    }
+
+    /// Tell this module where to look for the input addr
+    pub fn set_input_addr(&mut self, addr: Option<GuestAddr>) {
+        self.input_addr = addr;
+    }
+}
+
+impl<I, S> EmulatorModule<I, S> for RedirectStdinModule
+where
+    I: Unpin + HasLen + Debug,
+    S: Unpin,
+{
+    fn first_exec<ET>(
+        &mut self,
+        _qemu: Qemu,
+        emulator_modules: &mut EmulatorModules<ET, I, S>,
+        _state: &mut S,
+    ) where
+        ET: EmulatorModuleTuple<I, S>,
+    {
+        emulator_modules.pre_syscalls(Hook::Function(syscall_read_hook::<ET, I, S>));
+    }
+
+    fn pre_exec<ET>(
+        &mut self,
+        _qemu: Qemu,
+        _emulator_modules: &mut EmulatorModules<ET, I, S>,
+        _state: &mut S,
+        input: &I,
+    ) where
+        ET: EmulatorModuleTuple<I, S>,
+    {
+        assert!(self.input_addr.is_some());
```

Comment:
this won't be `some` if the user uses `set_input_addr` from the harness, right?

Reply:
No. Or use it with with_input_addr() constructor